src/hooks/usePricing.ts

```diff
@@ -0,0 +1,89 @@
+import { currencyService } from '@/components/services/currency.service';
+import { ProductsProps, stripeService } from '@/components/services/stripe.service';
+import { useEffect, useReducer } from 'react';
+import { CouponType } from '@/pages/api/stripe/get_coupons';
+
+interface UseStripeProductsAndCurrencyResponse {
+  products: ProductsProps | undefined;
+  loadingCards: boolean;
+  currency: string;
+  coupon: CouponType | null;
+}
+
+type ActionType =
+  | { type: 'SET_PRODUCTS'; payload: ProductsProps | undefined }
+  | { type: 'SET_LOADING_CARDS'; payload: boolean }
+  | { type: 'SET_CURRENCY'; payload: string }
+  | { type: 'SET_COUPON'; payload: CouponType | null };
+
+const reducer = (state: any, action: ActionType) => {
+  switch (action.type) {
+    case 'SET_PRODUCTS':
+      return { ...state, products: action.payload };
+    case 'SET_LOADING_CARDS':
+      return { ...state, loadingCards: action.payload };
+    case 'SET_CURRENCY':
+      return { ...state, currency: action.payload };
+    case 'SET_COUPON':
+      return { ...state, coupon: action.payload };
+    default:
+      return state;
+  }
+};
+
+function usePricing(couponCode?: CouponType): UseStripeProductsAndCurrencyResponse {
```

Comment:
I recommend you to use objects for parameters in the hooks, let me explain:

Imagine you are working with this hook, and tomorrow you have a new requirement:
"We need to allow users to switch the currency from the pricing page with a selector"

Now you will need to refactor this component, to accept a new parameter to accept the selected currency. Instead you can do this:

```
type UsePricingOptions = {
    couponCode: CouponType,
    currency: Currency
}

function usePricing(options: UsePricingOptions = {})...
```